### DISCUSSION TITLE
Linux: using 'brew shellenv' brokes sudo

### DISCUSSION THREAD

**Discussion by mt190502** [Homebrew/discussions#5672] [Linux] [answered]
### Additional Info

- [ ] I am using Windows Subsystem for Linux

### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.4.1
ORIGIN: https://github.com/Homebrew/brew
HEAD: 70672606c6ac07a5e8f75abeda7b8736e8285dbe
Last commit: 5 hours ago
Core tap JSON: 14 Oct 11:08 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 3.3.5 => /usr/bin/ruby
CPU: dodeca-core 64-bit unknown_0x6_0xa7
Clang: 19.1.0
Git: 2.47.0 => /bin/git
Curl: 8.9.1 => /bin/curl
Kernel: Linux 6.11.3-300.fc41.x86_64 x86_64 GNU/Linux
OS: Fedora release 41 (Forty One) (FortyOne)
Host glibc: 2.40
/usr/bin/gcc: 14.2.1
/usr/bin/ruby: 3.3.5
glibc: N/A
gcc@11: N/A
gcc: N/A
xorg: N/A
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

Hello.

Today I installed brew on my linux system. As a last step I added the following line in the fish config to activate shellenv.

/home/linuxbrew/.linuxbrew/bin/brew shellenv | source

After doing this, whenever I use sudo it asks for a password (as if sudo timeout was off). But if I remove the line above, sudo works fine. If I run it directly in fish shell without adding the above line to fish config, sudo works fine again.

I checked #4414 but it's not my answer

**Comment by SMillerDev** [comment]
> As a last step I added the following line in the fish config to activate shellenv.
> ```sh
> /home/linuxbrew/.linuxbrew/bin/brew shellenv | source
> ```

When is this activated?

**Reply by mt190502** [reply]
Yes. I checked env variables. HOMEBREW_ variables are available.

**Comment by gromgit** [comment, marked as answer]
> I checked #4414 but it's not my answer

True, it's almost certainly https://github.com/Homebrew/brew/pull/17694 instead.

To resolve this, read the comment thread starting from https://github.com/Homebrew/brew/issues/17915#issuecomment-2283157831, especially my comments to each proposed workaround.

**Reply by mt190502** [reply]
Oh, I'm sorry. I overlooked those problems. 
I wrote a callback script that calls the 'script' command. And I added 'brew shellenv' output to manually in fish config. Just worked

**Reply by gromgit** [reply]
> And I added 'brew shellenv' output to manually in fish config

Why not just wrap the `brew shellenv` invocation in `script` instead? Hardcoding current `brew shellenv` output means you won't get future bug fixes and improvements.

**Reply by mt190502** [reply]
> Why not just wrap the `brew shellenv` invocation in `script` instead? Hardcoding current `brew shellenv` output means you won't get future bug fixes and improvements.

When I use 'brew shellenv' with callback script, this causes an endless loop (I tested with empty fish config)
When I use without callback script, brew brokes sudo

![image](https://github.com/user-attachments/assets/78d95d0b-3fc1-4dc4-8811-9d676639aa67)

**Reply by gromgit** [reply]
Why are you using a callback script instead of a normal wrapper script, and what does your callback script look like?

**Reply by mt190502** [reply]
I may have used the wrong term

```bash
#!/bin/bash
faketty() {
  script -qefc "$(printf "%q " "$@")" /dev/null
}

faketty "/home/linuxbrew/.linuxbrew/bin/brew" "$@"
```

**Reply by gromgit** [reply]
And what's the output of `ls -l /home/linuxbrew/.linuxbrew/bin/brew`?

**Reply by mt190502** [reply]
> And what's the output of `ls -l /home/linuxbrew/.linuxbrew/bin/brew`?

Original brew script file. I don't modified any file under /home/linuxbrew

```
❯ ls -l /home/linuxbrew/.linuxbrew/bin/brew
lrwxrwxrwx. fedora fedora 20 B Mon Oct 14 14:08:11 2024  /home/linuxbrew/.linuxbrew/bin/brew ⇒ ../Homebrew/bin/brew
```

**Reply by gromgit** [reply]
If you rename your wrapper script to `brew.new`, then run `brew.new shellenv`, does it still loop infinitely?

**Reply by mt190502** [reply]
> If you rename your wrapper script to `brew.new`, then run `brew.new shellenv`, does it still loop infinitely?

Yes. The plot unfolds exactly like this (I guess)

- Fish shell calls a special script: `/path/to/script/brew shellenv`
- Custom script calls original brew with faketty function
- brew runs fish shell
- Fish shell calls the custom script again

**Reply by mt190502** [reply]
Do you have any idea?

**Reply by gromgit** [reply]
I don't use `fish`, but you might want to try tracing your `fish`'s login sequence, something like this:
```
fish_trace=1 fish -l -o /tmp/fish_login.log
```
then look in `/tmp/fish_login.log` to see if it tells you what's going on.

**Reply by mt190502** [reply]
https://paste.centos.org/view/83e3d738
Stucks on 'brew shellenv' step

**Reply by gromgit** [reply]
That log file looks incomplete, perhaps because the login shell spawned is still running. Try typing `exit`, then check the log file again.

**Reply by mt190502** [reply]
I open two terminals. After editing config in the first terminal, I type the command you mentioned in the second terminal.

**Reply by gromgit** [reply]
OK, close the second terminal, open a new one, then run this command:
```
fish_trace=1 fish -l -o /tmp/fish_login.log -c exit
```
That should spawn a login shell, run through your entire login sequence, then automatically exit. Check the log file after that.

**Reply by mt190502** [reply]
A similar situation occurred. Fish shell could not execute `exit` command

**Reply by gromgit** [reply]
just to be sure, what does `cat /home/fedora/.config/fish/config.fish` output?

**Reply by gromgit** [reply]
> Fish shell could not execute `exit` command

What do you see that tells you "`exit` cannot be executed"?

**Reply by mt190502** [reply]
> just to be sure, what does `cat /home/fedora/.config/fish/config.fish` output?

```
❯ cat .config/fish/config.fish | grep -v '^#'
$HOME/scripts/brew shellenv | source
```

**Reply by mt190502** [reply]
> > Fish shell could not execute `exit` command
> 
> What do you see that tells you "`exit` cannot be executed"?

![image](https://github.com/user-attachments/assets/efb263c6-56b2-43e6-9fd7-bf90ab9d67a8)

**Reply by gromgit** [reply]
>```
>❯ cat .config/fish/config.fish | grep -v '^#'
>$HOME/scripts/brew shellenv | source
>```

Change that line to:
```
$HOME/scripts/brew shellenv | tee -a /tmp/brew_shellenv.out | source
```
then open a new terminal and see what `/tmp/brew_shellenv.out` contains.

**Reply by mt190502** [reply]
```
❯ cat /tmp/brew_shellenv.out
^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^C/bin/bash: warning: shell level (1000) too high, resetting to 1
```

**Reply by gromgit** [reply]
And what are the exact contents of `$HOME/scripts/brew` contain?

**Reply by mt190502** [reply]
> And what are the exact contents of `$HOME/scripts/brew` contain?

```
❯ cat scripts/brew
#!/bin/bash
faketty() {
  script -qefc "$(printf "%q " "$@")" /dev/null
}

faketty "/home/linuxbrew/.linuxbrew/bin/brew" "$@"
```
